DEV-413 special DFX voting

ThaBrad opened the pull request:
Static Voting item for the special DFX voting

**--> end-date probably has to change** 

currently = `endDate: "2022-07-30T23:59:59.000Z"`

Will not be listed, at CFP page, after this date.

Krysh90 reviewed (approved):
Two short questions:
1. Why is this special CFP voting hard-coded in the app? Wouldn't it make more sense, to retrieve this information via API?
2. If it is hardcoded, what about translations?

ThaBrad commented:
@Krysh90 
1. Because API simply fetches the stuff from GitHub and "forwards" it. No Persistence.
2. Only got the info now for that => added

ThaBrad commented:
@davidleomay 
I still have no info about start and end date => currently hard-coded to
```
    startDate: "2022-07-23T23:59:59.000Z",
    endDate: "2022-07-30T23:59:59.000Z"
```

ThaBrad commented:
Btw, I changed the staking balance check (and also updated the userDetailsDto).

From:
`setCanVote(stakingRoutes.find((r) => r.balance >= 100) != null);
`To:
`setCanVote(userDetail.stakingBalance >= 100);`

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+
+export default withSession(SpecialVotingScreen);
+
+const description1 = `
+Jeder Nutzer des DFI Stakingservice von DFX erhält eine gewichtete Stimme auf Basis der investierten DFI. Zur Wahl stehen zwei alternative Abstimmungsmechanismen. 
+
+
+Auswahlmöglichkeiten:
+
+DFX soll mit allen Masternodes als Kollektiv mit ja/nein/neutral abstimmen
+DFX soll mit allen Masternodes anteilig der abgegebenen Stimmen mit ja/nein/neutral abstimmen
+Neutral
+`
+
+const description2 = `
+Kollektivbetrachtung:
+
+Abhängig der Wahlergebnisse der jeweiligen CFPs/DFIPs von der DFX Community stimmt DFX mit allen Masternodes für seine Kunden ab. 
+> 50% - DFX votet mit allen Masternodes „nein“
+50% - DFX votet mit allen Masternodes „neutral“
+>50% - DFX votet mit allen Masternodes „ja“
+Die Kollektivbetrachtung wird nur angewendet, so lange DFX <10 % der DeFiChain Masternodes für seine Kunden betreibt. Wird dieser Schwellenwert überschritten wird DFX automatisch die anteilige Betrachtung wählen
+
+
+Anteilige Betrachtung:
+
+Abhängig der Wahlergebnisse der jeweiligen CFPs von der DFX Community stimmt DFX für seine Kunden anteilig der abgegeben Stimmen mit ja/nein/neutral ab
+
+
+`
```
Unused variables

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+import React, { useEffect, useState } from "react";
+import { useTranslation } from "react-i18next";
+import { StyleSheet, View } from "react-native";
+import { StackedBarChart } from "react-native-chart-kit";
+import { AbstractChartConfig } from "react-native-chart-kit/dist/AbstractChart";
+import { StackedBarChartData } from "react-native-chart-kit/dist/StackedBarChart";
+import { DataTable, Text } from "react-native-paper";
+import AppLayout from "../components/AppLayout";
+import Loading from "../components/util/Loading";
+import Colors from "../config/Colors";
+import { DeFiButton } from "../elements/Buttons";
+import { RadioButton } from "../elements/RadioButton";
+import { SpacerV } from "../elements/Spacers";
+import { CompactRow, CompactCell } from "../elements/Tables";
+import { H1, H3, T5 } from "../elements/Texts";
+import withSession from "../hocs/withSession";
+import useAuthGuard from "../hooks/useAuthGuard";
+import { CfpResult, ResultStatus, VotingType } from "../models/CfpResult";
+import { CfpVote, CfpVotes } from "../models/User";
+import { getSettings, getUserDetail, putCfpVotes } from "../services/ApiService";
+import { Session } from "../services/AuthService";
+import NotificationService from "../services/NotificationService";
+import AppStyles from "../styles/AppStyles";
```
Lot's of unused imports

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  // const getData = (result: CfpResult): StackedBarChartData => {
+  //   return {
+  //     labels: [],
+  //     legend: [
+  //       `${t("cfp.yes")} (${Math.round((result.totalVotes.yes / result.totalVotes.total) * 100)}%)`,
+  //       `${t("cfp.neutral")} (${Math.round((result.totalVotes.neutral / result.totalVotes.total) * 100)}%)`,
+  //       `${t("cfp.no")} (${Math.round((result.totalVotes.no / result.totalVotes.total) * 100)}%)`,
+  //     ],
+  //     data: [[], [result.totalVotes.yes, result.totalVotes.neutral, result.totalVotes.no]],
+  //     barColors: [Colors.Success, Colors.LightGrey, Colors.Error],
+  //   };
+  // };
```
Remove

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+    });
+  };
+
+  const config: AbstractChartConfig = {
+    backgroundColor: Colors.Blue,
+    backgroundGradientFrom: Colors.Blue,
+    backgroundGradientTo: Colors.Blue,
+    decimalPlaces: 0,
+    color: (opacity = 1) => `rgba(255,  255,  255,  ${opacity})`,
+    labelColor: (opacity = 1) => `rgba(255,  255,  255,  ${opacity})`,
+    propsForBackgroundLines: {
+      stroke: Colors.Blue,
+    },
+  };
```
Unused variable

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+        }
+
+        setCfpResults(results);
+        setVotes(votes);
```
Useless line `setVotes(votes)`

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+  const [cfpFilter, setCfpFilter] = useState<"all" | "dfx">("all");
+  const [cfpResults, setCfpResults] = useState<CfpResult[]>();
+  const [canVote, setCanVote] = useState(false);
+  const [isVotingOpen, setIsVotingOpen] = useState(false);
```
Unused state `isVotingOpen`

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+          <Loading size="large" />
+        ) : (
+          <>
+            {DfxCfpNumbers.length > 0 && (
+              <>
+                <View style={AppStyles.containerHorizontalWrap}>
+                  <RadioButton
+                    label={t("cfp.dfx_only")}
+                    onPress={() => setCfpFilter("dfx")}
+                    checked={cfpFilter === "dfx"}
+                  />
+                  <RadioButton label={t("cfp.all")} onPress={() => setCfpFilter("all")} checked={cfpFilter === "all"} />
+                </View>
+                <SpacerV height={50} />
+              </>
+            )}
```
Unused code

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+import NotificationService from "../services/NotificationService";
+import AppStyles from "../styles/AppStyles";
+
+const DfxCfpNumbers: number[] = [];
```
Unused variable

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+const SpecialVotingScreen = ({ session }: { session?: Session }) => {
+  const { t } = useTranslation();
+  const [isLoading, setIsLoading] = useState(true);
+  const [cfpFilter, setCfpFilter] = useState<"all" | "dfx">("all");
```
Unused state `cfpFilter`

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+                    <T5 text={t("specialVoting.explanation") + "\n\n\n"} /* style={AppStyles.center}  */ />
+                  </View>
+
+                  {(canVote || isSpecialVoting(result)) && (
```
This is `SpecialVotingScreen`, `isSpecialVoting` is always true.

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+              .map((result) => (
+                <View key={result.number} style={{ width: "100%" }}>
+                  <View style={AppStyles.halfWidth} >
+                    <T5 text={t("specialVoting.explanation") + "\n\n\n"} /* style={AppStyles.center}  */ />
```
Remove commented code

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+                  )}
+
+                  <View style={AppStyles.halfWidth} >
+                    <T5 text={"\n\n\n" + t("specialVoting.description")} /* style={AppStyles.center}  */ />
```
Remove commented code

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,266 @@
+  );
+};
+
+const styles = StyleSheet.create({
+  cfpContainer: {
+    flexDirection: "row",
+    flexWrap: "wrap-reverse",
+    justifyContent: "center",
+    alignItems: "center",
+  },
+  chartContainer: {
+    width: 250,
+    height: 180,
+    alignItems: "flex-end",
+    overflow: "hidden",
+  },
+  radioRow: {
+    flexDirection: "row",
+    alignItems: "center",
+    justifyContent: "space-between",
+    paddingHorizontal: 16,
+  },
+  voteContainer: {
+    justifyContent: "center",
+  },
+});
```
Unused styles

Krysh90 commented on app/i18n/pt.json:
```diff
@@ -339,6 +348,15 @@
+      "collectively": "",
+      "proportionally": "",
+      "explanation": "",
+      "description": ""
```
Empty values for pt?

ThaBrad replied:
uff… I need to make that pre-comit script
But in any case, pt is not an officially supported language.

davidleomay replied:
Please add the PT translations before merging.

Krysh90 commented on app/components/HeaderContent.tsx:
```diff
@@ -58,6 +58,9 @@ const HeaderContent = ({ session, settings }: { session?: Session; settings?: Ap
           </DeFiButton>
+          <DeFiButton onPress={() => navigate(Routes.SpecialVoting)} style={styles.button} compact>
+            {t("specialVoting.link")}
+          </DeFiButton>
```
Just curious - will this button get removed again or at least hidden?  Or do we want to always show special votings even if there are no special votings active?

ThaBrad replied:
We'll remove it manually after a certain while. Currently we'll continue to show it to be transparent. 
We might use it thogh for future internal votings.

davidleomay commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -0,0 +1,163 @@
+              ?.sort((a, b) => a.number - b.number)
+              .map((result) => (
+                <View key={result.number} style={{ width: "100%" }}>
+                  <View style={AppStyles.halfWidth} >
```
Half-width looks IMO a bit weird on smaller screens:
![image](https://user-images.githubusercontent.com/85513542/181038157-5941a105-4a25-4888-935d-28ef884875e5.png)
I would propose to go with 90%.

davidleomay replied:
There's a bug related to the halfWidth style. The background at the bottom is not blue:
![image](https://user-images.githubusercontent.com/85513542/181038454-6fa03707-aca6-415a-bd47-e5bef13b05d6.png)

davidleomay reviewed (commented):
Oli reported typos in the german translation. Use the following text:

---

Abstimmung: DFX Masternode CFP/DFIP Bewertungslogik

Jeder Nutzer des DFI Stakingservice von DFX erhält eine gewichtete Stimme auf Basis der investierten DFI. Zur Wahl stehen zwei alternative Abstimmungsmechanismen. Auswahlmöglichkeiten:

- Kollektiv: DFX soll mit allen Masternodes als Kollektiv mit ja/nein/neutral abstimmen
- Anteilig: DFX soll mit allen Masternodes anteilig der abgegebenen Stimmen mit ja/nein/neutral abstimmen

*** voting here ***

Kollektivbetrachtung: 
Abhängig der Wahlergebnisse der jeweiligen CFPs/DFIPs von der DFX Community stimmt DFX mit allen Masternodes für seine Kunden ab. 
< 50% - DFX votet mit allen Masternodes „nein“
50% - DFX votet mit allen Masternodes „neutral“
>50% - DFX votet mit allen Masternodes „ja“
Die Kollektivbetrachtung wird nur angewendet, so lange DFX <10 % der DeFiChain Masternodes für seine Kunden betreibt. Wird dieser Schwellenwert überschritten wird DFX automatisch die anteilige Betrachtung wählen

Anteilige Betrachtung:
Abhängig der Wahlergebnisse der jeweiligen CFPs von der DFX Community stimmt DFX für seine Kunden anteilig der abgegeben Stimmen mit ja/nein/neutral ab

davidleomay reviewed (commented):
![image](https://user-images.githubusercontent.com/85513542/181054692-d7a16b0a-7f0b-48be-be59-4d9684cd8881.png)
Title is duplicated.

davidleomay reviewed (commented):
English text below

---

Voting: DFX Masternode CFP/DFIP Voting Logic.

Each user of DFX's DFI Staking Service receives a weighted vote based on their DFI invested. There are two alternative voting mechanisms to choose from. Choices:

- Collective: DFX shall vote yes or no with all masternodes as a collective.
- Proportional: DFX shall vote yes or no with all masternodes proportionate to the votes cast.

Your vote
Collective
Proportional
No preference



Collective View: 
Depending on the voting results of the respective CFPs/DFIPs from the DFX community, DFX votes with all masternodes for its customers. 

< 50% - DFX votes "no" with all masternodes.
>= 50% - DFX votes "yes" with all masternodes

Collective consideration is only applied as long as DFX operates <10% of DeFiChain masternodes for its customers. If this threshold is exceeded, DFX will automatically select proportional consideration

Proportional Consideration:
Depending on the voting results of the respective CFPs/DFIPs from the DFX community, DFX votes yes/no/neutral for its customers proportionally to the votes cast

Krysh90 commented on app/screens/SpecialVotingScreen.tsx:
```diff
@@ -110,39 +112,39 @@ const SpecialVotingScreen = ({ session }: { session?: Session }) => {
-                  <View style={AppStyles.halfWidth} >
-                    <T5 text={"\n\n\n" + t("specialVoting.description")} />
-                  </View>
+                  <T5 text={t("specialVoting.explanation") + "\n\n\n"} />
```
just curious - why using 3x new lines instead of `<SpacerV height={x} />`?

ThaBrad replied:
- No new View in ViewHierarchy to render/compute
- Wanted to keep the relative line height for the paragraph
